kani-driver/src/exe_trace.rs

```diff
@@ -0,0 +1,264 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+use crate::session::KaniSession;
+use kani_metadata::HarnessMetadata;
+use std::collections::hash_map::DefaultHasher;
+use std::ffi::OsString;
+use std::fs;
+use std::hash::{Hash, Hasher};
+use std::path::Path;
+use std::process::Command;
+
+/// Module for parsing deterministic values from CBMC output traces,
+/// generating executable traces, and adding them to the user's source code.
+
+impl KaniSession {
+    /// The main driver for generating executable traces and adding them to source code.
+    pub fn exe_trace_main(&self, specialized_obj: &Path, harness: &HarnessMetadata) {
+        if self.args.gen_exe_trace {
+            let (det_vals, interp_det_vals) = parser::get_det_vals(specialized_obj);
+            let (exe_trace, exe_trace_func_name) =
+                format_unit_test(&harness.mangled_name, &det_vals[..], &interp_det_vals[..]);
+
+            println!("Executable trace:\n");
+            println!("{}", exe_trace);
+
+            if self.args.add_exe_trace_to_src {
+                if !self.args.quiet {
+                    println!("Now modifying the source code to include the executable trace.");
+                }
+                let proof_harness_end_line: usize = harness
+                    .original_end_line
+                    .parse()
+                    .expect("Couldn't convert proof harness line from str to usize");
+                self.modify_src_code(
+                    &harness.original_file,
+                    proof_harness_end_line,
+                    &exe_trace,
+                    &exe_trace_func_name,
+                );
+            } else {
+                println!(
+                    "To automatically add this executable trace to the src code, run Kani with `--add-exe-trace-to-src`."
+                );
+            }
+        }
+    }
+
+    /// Add the exe trace to the user's source code, format it, and short circuit if code already present.
+    fn modify_src_code(
+        &self,
+        src_file_path: &str,
+        proof_harness_end_line: usize,
+        exe_trace: &str,
+        exe_trace_func_name: &str,
+    ) {
+        // Prep the source code and exec trace func.
+        let src_as_str =
```

Comment:
Can you please use `read_lines()` instead? https://doc.rust-lang.org/rust-by-example/std_misc/file/read_lines.html

Reply:
I think `read_lines()` would've been better with the previous logic (the logic this comment was based on), since it did a lot of conversions between a) the file as a full string and b) the file as a vector of lines. However, I refactored the logic to work only on the full string, and I think the new solution is cleaner and the same efficiency as a line-by-line approach with `read_lines`. I'd be happy to discuss this further, though, if you think it should be improved.

Reply:
TODO: remove 2nd string allocation

Reply:
Done